File: energy.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* Constants for numerical stability */
#define EPSILON_SING 1e-12

// Compute Lennard-Jones potential energy for distance r.
double lennard_jones_potential(double r, double epsilon, double sigma) {
    if (r < EPSILON_SING) {
        return 1e12;  // Avoid singularity
    }
    double sr6 = pow(sigma / r, 6);
    return 4.0 * epsilon * (sr6 * sr6 - sr6);
}

// Compute harmonic bond potential energy for a bond of length r.
double bond_potential(double r, double b, double k_b) {
    return k_b * (r - b) * (r - b);
}

// Compute total energy given positions (flattened array of length n_beads*3).
double total_energy(const double *positions, int n_beads, 
                    double epsilon, double sigma, double b, double k_b) {
    double energy = 0.0;
    int i, j;
    // Bond energy: over consecutive beads.
    for (i = 0; i < n_beads - 1; i++) {
        double dx = positions[(i+1)*3 + 0] - positions[i*3 + 0];
        double dy = positions[(i+1)*3 + 1] - positions[i*3 + 1];
        double dz = positions[(i+1)*3 + 2] - positions[i*3 + 2];
        double r = sqrt(dx*dx + dy*dy + dz*dz);
        energy += bond_potential(r, b, k_b);
    }
    // Lennard-Jones energy: over all unique pairs.
    for (i = 0; i < n_beads; i++) {
        for (j = i+1; j < n_beads; j++) {
            double dx = positions[j*3 + 0] - positions[i*3 + 0];
            double dy = positions[j*3 + 1] - positions[i*3 + 1];
            double dz = positions[j*3 + 2] - positions[i*3 + 2];
            double r = sqrt(dx*dx + dy*dy + dz*dz);
            energy += lennard_jones_potential(r, epsilon, sigma);
        }
    }
    return energy;
}

/*
 * Compute the analytical gradient dE/dx (for all coordinates) and write the result into gradient.
 *
 * For the bond potential:
 *   V_bond = k_b * (r - b)^2,   with r = ||x_(i+1) - x_i||
 *   dV/dr = 2 * k_b * (r - b)
 *   Chain rule gives:
 *     dE/dx_i   = -2 * k_b * (r - b) * ( (x_{i+1} - x_i) / r )
 *     dE/dx_{i+1} = +2 * k_b * (r - b) * ( (x_{i+1} - x_i) / r )
 *
 * For the Lennard-Jones potential:
 *   V_LJ = 4 * epsilon * [ (sigma/r)^12 - (sigma/r)^6 ]
 *   dV/dr = -48 * epsilon * sigma^12 / r^13 + 24 * epsilon * sigma^6 / r^7
 *   Using dr/dx_i = -(x_j - x_i)/r, we get:
 *     dE/dx_i = 48 * epsilon * sigma^12 * (x_j - x_i) / r^14 
 *                - 24 * epsilon * sigma^6 * (x_j - x_i) / r^8,
 *   and the opposite for bead j.
 *
 * The gradient vector is stored as a flattened array of length n_beads*3.
 */
void compute_gradient(const double *positions, int n_beads, 
                      double epsilon, double sigma, double b, double k_b, 
                      double *gradient) {
    int i, j;
    // Initialize gradient to zero.
    for (i = 0; i < n_beads * 3; i++) {
        gradient[i] = 0.0;
    }
    
    // Bond potential gradient.
    for (i = 0; i < n_beads - 1; i++) {
        double dx = positions[(i+1)*3 + 0] - positions[i*3 + 0];
        double dy = positions[(i+1)*3 + 1] - positions[i*3 + 1];
        double dz = positions[(i+1)*3 + 2] - positions[i*3 + 2];
        double r = sqrt(dx*dx + dy*dy + dz*dz);
        if (r > EPSILON_SING) {
            double deriv = 2.0 * k_b * (r - b);  // dV/dr for bond
            double gx = deriv * (dx / r);
            double gy = deriv * (dy / r);
            double gz = deriv * (dz / r);
            // For energy: bead i gets -grad_contrib, bead i+1 gets +grad_contrib.
            gradient[i*3 + 0]   -= gx;
            gradient[i*3 + 1]   -= gy;
            gradient[i*3 + 2]   -= gz;
            gradient[(i+1)*3 + 0] += gx;
            gradient[(i+1)*3 + 1] += gy;
            gradient[(i+1)*3 + 2] += gz;
        }
    }
    
    // Lennard-Jones potential gradient.
    for (i = 0; i < n_beads; i++) {
        for (j = i+1; j < n_beads; j++) {
            double dx = positions[j*3 + 0] - positions[i*3 + 0];
            double dy = positions[j*3 + 1] - positions[i*3 + 1];
            double dz = positions[j*3 + 2] - positions[i*3 + 2];
            double r = sqrt(dx*dx + dy*dy + dz*dz);
            if (r > EPSILON_SING) {
                double term1 = 48.0 * epsilon * pow(sigma, 12) / pow(r, 13);
                double term2 = 24.0 * epsilon * pow(sigma, 6) / pow(r, 7);
                double deriv = -term1 + term2;  // dV/dr for Lennard-Jones.
                double gx = deriv * (dx / r);
                double gy = deriv * (dy / r);
                double gz = deriv * (dz / r);
                // For energy: bead i gets -grad_contrib and bead j gets +grad_contrib.
                gradient[i*3 + 0]   -= gx;
                gradient[i*3 + 1]   -= gy;
                gradient[i*3 + 2]   -= gz;
                gradient[j*3 + 0]   += gx;
                gradient[j*3 + 1]   += gy;
                gradient[j*3 + 2]   += gz;
            }
        }
    }
}

/*
 * Initialize a protein with n_beads arranged almost linearly in 3D.
 * The "fudge" parameter adds a small spiral perturbation in the y- and z-directions.
 *
 * This function mimics the following Python code:
 *
 *   def initialize_protein(n_beads, dimension=3, fudge=1e-5):
 *       positions = np.zeros((n_beads, dimension))
 *       for i in range(1, n_beads):
 *           positions[i, 0] = positions[i-1, 0] + 1
 *           positions[i, 1] = fudge * np.sin(i)
 *           positions[i, 2] = fudge * np.sin(i*i)
 *       return positions
 *
 * The positions array is a flattened array of length n_beads*3.
 */
void initialize_protein(double *positions, int n_beads, int dimension, double fudge) {
    int i, d;
    // Set all positions to zero.
    for (i = 0; i < n_beads * dimension; i++) {
        positions[i] = 0.0;
    }
    // Initialize positions for i >= 1.
    for (i = 1; i < n_beads; i++) {
        // x-coordinate: previous x + 1.0
        positions[i * dimension + 0] = positions[(i - 1) * dimension + 0] + 1.0;
        // y-coordinate: fudge * sin(i) (if dimension > 1)
        if (dimension > 1) {
            positions[i * dimension + 1] = fudge * sin((double)i);
        }
        // z-coordinate: fudge * sin(i*i) (if dimension > 2)
        if (dimension > 2) {
            positions[i * dimension + 2] = fudge * sin((double)(i * i));
        }
    }
}

#ifdef TEST_ENERGY
// Main function for testing.
int main(void) {
    int n_beads = 10;
    int dimension = 3;
    int n_coords = n_beads * dimension;
    double epsilon = 1.0, sigma = 1.0, b = 1.0, k_b = 100.0;
    double fudge = 1e-5;
    
    // Allocate positions and gradient.
    double *positions = (double *)malloc(n_coords * sizeof(double));
    double *grad = (double *)malloc(n_coords * sizeof(double));
    if (!positions || !grad) {
        fprintf(stderr, "Memory allocation failed.\n");
        return 1;
    }
    
    // Initialize positions reproducibly.
    initialize_protein(positions, n_beads, dimension, fudge);
    
    double energy = total_energy(positions, n_beads, epsilon, sigma, b, k_b);
    compute_gradient(positions, n_beads, epsilon, sigma, b, k_b, grad);
    
    printf("Total Energy: %e\n", energy);
    printf("Gradient:\n");
    for (int i = 0; i < n_coords; i++) {
        printf("%e ", grad[i]);
        if ((i+1) % dimension == 0) {
            printf("\n");
        }
    }
    
    free(positions);
    free(grad);
    return 0;
}
#endif

File: energy.py
import numpy as np

def lennard_jones_potential(r, epsilon=1.0, sigma=1.0):
    """
    Compute the Lennard-Jones potential energy for a given interparticle distance r.
    
    Parameters:
        r (float): Distance between two particles.
        epsilon (float): Depth of the potential well.
        sigma (float): Finite distance at which the potential is zero.
        
    Returns:
        float: Lennard-Jones potential energy.
    """
    if r < 1e-12:
        return 1e12  
    sr6 = (sigma / r) ** 6
    return 4 * epsilon * (sr6**2 - sr6)

def bond_potential(r, b=1.0, k_b=100.0):
    """
    Compute the harmonic bond potential energy for a bond.
    
    Parameters:
        r (float): Current bond length.
        b (float): Equilibrium bond length.
        k_b (float): Bond stiffness (force constant).
    
    Returns:
        float: Bond potential energy.
    """
    return k_b * (r - b) ** 2

def total_energy(positions, n_beads, epsilon=1.0, sigma=1.0, b=1.0, k_b=100.0):
    """
    Compute the total potential energy for a system of beads.
    
    Parameters:
        positions (np.ndarray): Flattened array of shape (n_beads*3,) representing bead positions.
        n_beads (int): Number of beads/particles.
        epsilon (float): Lennard-Jones parameter.
        sigma (float): Lennard-Jones parameter.
        b (float): Equilibrium bond length.
        k_b (float): Bond stiffness.
        
    Returns:
        float: Total potential energy.
    """
    energy = 0.0
    positions = positions.reshape(n_beads, 3)
    
    # Bond potential energy (adjacent beads).
    for i in range(n_beads - 1):
        r = np.linalg.norm(positions[i+1] - positions[i])
        energy += bond_potential(r, b, k_b)
    
    # Lennard-Jones potential energy (all unique pairs).
    for i in range(n_beads):
        for j in range(i + 1, n_beads):
            r = np.linalg.norm(positions[j] - positions[i])
            energy += lennard_jones_potential(r, epsilon, sigma)
    
    return energy

def compute_gradient(positions, n_beads, epsilon=1.0, sigma=1.0, b=1.0, k_b=100.0):
    """
    Compute the analytical gradient (dE/dx) of the total potential energy.
    
    For the bond potential:
        V_bond = k_b * (r - b)^2 with r = ||x_(i+1) - x_i||
        dV/dr = 2*k_b*(r - b) and dr/dx_i = -(x_{i+1} - x_i)/r,
        so that dE/dx_i = -2*k_b*(r - b)*(x_{i+1}-x_i)/r,
        and dE/dx_(i+1) = 2*k_b*(r - b)*(x_{i+1}-x_i)/r.
    
    For the Lennard-Jones potential:
        V_LJ = 4*epsilon*((sigma/r)^12 - (sigma/r)^6),
        dV/dr = -48*epsilon*sigma^12/r^13 + 24*epsilon*sigma^6/r^7,
        with dr/dx_i = -(x_j - x_i)/r, yielding
        dE/dx_i = 48*epsilon*sigma^12*(x_j-x_i)/r^14 - 24*epsilon*sigma^6*(x_j-x_i)/r^8.
    
    Parameters:
        positions (np.ndarray): Flattened array of shape (n_beads*3,).
        n_beads (int): Number of beads.
        epsilon, sigma: Lennard-Jones parameters.
        b, k_b: Bond potential parameters.
        
    Returns:
        np.ndarray: Flattened gradient array (dE/dx) of shape (n_beads*3,).
    """
    positions = positions.reshape(n_beads, 3)
    gradient = np.zeros_like(positions)
    
    # Bond potential gradient.
    for i in range(n_beads - 1):
        r_vec = positions[i+1] - positions[i]  # vector from bead i to bead i+1.
        r = np.linalg.norm(r_vec)
        if r > 1e-12:
            deriv = 2 * k_b * (r - b)  # dV/dr.
            grad_contrib = deriv * (r_vec / r)  # chain rule.
            # dE/dx_i = -grad_contrib, dE/dx_(i+1) = +grad_contrib.
            gradient[i]   -= grad_contrib
            gradient[i+1] += grad_contrib
    
    # Lennard-Jones potential gradient.
    for i in range(n_beads):
        for j in range(i + 1, n_beads):
            r_vec = positions[j] - positions[i]
            r = np.linalg.norm(r_vec)
            if r > 1e-12:
                deriv = -48 * epsilon * (sigma**12) / (r**13) + 24 * epsilon * (sigma**6) / (r**7)
                grad_contrib = deriv * (r_vec / r)
                # Change here: for energy we assign:
                # dE/dx_i = -grad_contrib and dE/dx_j = +grad_contrib.
                gradient[i]   -= grad_contrib
                gradient[j]   += grad_contrib
    
    return gradient.flatten()

def numerical_gradient(f, x, n_beads, epsilon=1.0, sigma=1.0, b=1.0, k_b=100.0, h=1e-5):
    """
    Compute the numerical gradient using central finite differences.
    
    Approximates dE/dx as:
        (E(x + h) - E(x - h)) / (2h)
    
    Parameters:
        f (callable): Energy function.
        x (np.ndarray): Flattened positions.
        n_beads (int): Number of beads.
        epsilon, sigma, b, k_b: Energy parameters.
        h (float): Step size.
    
    Returns:
        np.ndarray: Numerical gradient as a flattened array.
    """
    grad = np.zeros_like(x)
    for i in range(len(x)):
        x_forward = x.copy()
        x_backward = x.copy()
        x_forward[i] += h
        x_backward[i] -= h
        grad[i] = (f(x_forward, n_beads, epsilon, sigma, b, k_b) -
                   f(x_backward, n_beads, epsilon, sigma, b, k_b)) / (2 * h)
    return grad
